spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jdbc/DataSourceConfiguration.java

```diff
@@ -114,12 +118,39 @@ HikariDataSource dataSource(DataSourceProperties properties) {
 	 */
 	@Configuration(proxyBeanMethods = false)
 	@ConditionalOnMissingBean(DataSource.class)
-	@ConditionalOnProperty(name = "spring.datasource.type")
+	@Conditional(GenericCondition.class)
 	static class Generic {
 
 		@Bean
 		DataSource dataSource(DataSourceProperties properties) {
-			return properties.initializeDataSourceBuilder().build();
+			return properties.initializeDataSourceBuilder().type(determineType(properties)).build();
+		}
+
+		private static Class<? extends DataSource> determineType(DataSourceProperties properties) {
+			Class<? extends DataSource> type = properties.getType();
+			if (type == null) {
+				type = DataSourceBuilder.findType(properties.getClassLoader());
+			}
+			return (type != null) ? type : SimpleDriverDataSource.class;
+		}
+
+	}
+
+	static class GenericCondition extends AnyNestedCondition {
```

Comment:
We should rename this when merging. Something like `ExplicitDataSourceTypeOrUrlCondition`.